java/test/behaviour/util/Util.java

```diff
@@ -43,4 +51,43 @@ public static void assertThrowsWithMessage(Runnable function, String message) {
             assert e.toString().toLowerCase().contains(message.toLowerCase());
         }
     }
+
+    public static boolean JSONListMatches(List<JSON> lhs, List<JSON> rhs) {
+            if (lhs.size() != rhs.size()) return false;
+            Set<Integer> rhsMatches = new HashSet<>();
+            for (JSON lhsItem: lhs) {
+                for (int i = 0; i < rhs.size(); i++) {
+                    if (rhsMatches.contains(i)) continue;
+                    JSON rhsItem = rhs.get(i);
+                    if (JSONMatches(lhsItem, rhsItem)) {
+                        rhsMatches.add(i);
+                        break;
+                    }
+                }
+            }
+            return rhsMatches.size() == rhs.size();
+    }
+
+    public static boolean JSONMatches(JSON lhs, JSON rhs) {
+        if (lhs.isObject()) {
+            if (!rhs.isObject()) return false;
+            Map<String, JSON> lhsMap = lhs.asObject();
+            Map<String, JSON> rhsMap = rhs.asObject();
+            if (lhsMap.size() != rhsMap.size()) return false;
+            return lhsMap.keySet().stream().allMatch(key -> JSONMatches(lhsMap.get(key), rhsMap.get(key)));
+        } else if (lhs.isArray()) {
+            if (!rhs.isArray()) return false;
+            return JSONListMatches(lhs.asArray(), rhs.asArray());
+        } else if (lhs.isNumber()) {
```

Comment:
I believe the non-object and non-array matches can simplfy into `a.equals(b)` using the JSON equality defined on primitives that you've written